### PR TITLE
shim: Add option to read and print the logs from the agent

### LINKED ISSUE #171: Add option to print the logs from the agent
# Description of problem

It's really difficult to debug the agent when `use_vsock` is enabled, since the logs from the agent are ignored because there is no proxy to read and print them. Kata shim should have an option to print the agent logs in order to make debugging easier.

### PR DESCRIPTION
**Pull request by devimc**
shim is now able to read and print the agent logs, making the
debugging processes easier when `use_vsocks` is true.

fixes #171

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -97,15 +101,81 @@ func setThreads() {
...
+	}
+	if addr.Scheme != "" && addr.Scheme != "unix" {
+		return "", errors.New("invalid address scheme")
+	}
+	return addr.Host + addr.Path, nil
+}
+
+func printAgentLogs(sock string) error {
+	// Don't return an error if nothing has been provided.
+	// This flag is optional.
```
Comment (grahamwhaley) [inline main.go]: 'not' optional? the comment does not seem to align with how the code reads?

File: main.go [hunk 2]
```diff
@@ -97,15 +101,81 @@ func setThreads() {
...
+		return nil
+	}
+
+	agentLogsAddr, err := unixAddr(sock)
+	if err != nil {
+		logger().WithField("socket-address", sock).WithError(err).Fatal("invalid agent logs socket address")
+		return err
+	}
+
+	// Check permissions socket for "other" is 0.
```
Comment (grahamwhaley) [inline main.go]: Is this a security check? If so, can you make that clear in the comment, just to make it nice and clear for the future :-)

File: main.go (hunk 1 above)
Comment (devimc) [reply main.go]: you're right

File: main.go (hunk 2 above)
Comment (devimc) [reply main.go]: ok :smile:

**Comment by devimc** [comment]
@sboeuf 

> But the question is, from a kata-runtime perspective, which shim are you going to pick to listen onto the socket?

the fist shim (sandbox)


> My concern is that you will have each container level shim listening onto this socket, while we want only a single entity listening to this (since the agent is considered pod level), right?

yes, only the sandbox shim is going to listen it, see https://github.com/devimc/kata-runtime/commit/09b5bb94eed7a8f57a76af69f2b7d16821dbbbbb

**Comment by devimc** [comment]
@grahamwhaley  changes applied, thanks

**Comment by sboeuf** [comment]
@devimc 
>yes, only the sandbox shim is going to listen it, see devimc/kata-runtime@09b5bb9

Ok this sounds good, but are you sure the `ContainerType` annotation is going to be provided in case of Docker? This will work for CRI callers (CRI-O, CRI/containerd), but I am worried this will not work for Docker.

**Comment by devimc** [comment]
@sboeuf it's working on docker and the CI seems happy https://github.com/kata-containers/runtime/pull/1597

**Comment by grahamwhaley** [comment]
Fedora CI failed - I'll nudge: 

```
state container 
  with workload [true], timeWait 5
  /tmp/jenkins/workspace/kata-containers-shim-fedora-PR/go/src/github.com/kata-containers/tests/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:46
Running command '/usr/local/bin/kata-runtime [kata-runtime --log=/tmp/bundle345590837/log run --bundle=/tmp/bundle345590837 --console= --pid-file=/tmp/bundle345590837/pid --detach ahkdkPuqiwbRLh9nfCIU]'
command failed error 'exit status 1'
[kata-runtime --log=/tmp/bundle345590837/log run --bundle=/tmp/bundle345590837 --console= --pid-file=/tmp/bundle345590837/pid --detach ahkdkPuqiwbRLh9nfCIU]
Timeout: 120 seconds
Exit Code: 1
Stdout: 
Stderr: Failed to check if grpc server is working: rpc error: code = Unavailable desc = transport is closing

Running command '/usr/local/bin/kata-runtime [kata-runtime list --quiet]'
[kata-runtime list --quiet]
Timeout: 120 seconds
Exit Code: 0
Stdout: 
Stderr: 
```

Ignoring Power8 CI for now until I hear it is back to being stable.

**Pull request merged**